docs(schema): create Working with Joins page under Fundamentals section (#3276)
cube-js/cube.js: docs/content/Schema/Fundamentals/Working-with-Joins.md

---
title: Working with Joins
permalink: /schema/fundamentals/joins
category: Data Schema
subCategory: Fundamentals
menuOrder: 4
redirect_from:
  - /direction-of-joins
  - /many-to-many-relationship
---

## Many-to-many joins

A many-to-many relationship occurs when multiple records in a cube are
associated with multiple records in another cube.

For example, let's say we have two cubes, `Topics` and `Posts`, pointing to the
`topics` and `posts` tables in our database respectively. A `Post` can have more
than one `Topic`, and a `Topic` may have more than one `Post`.

In a database, you would most likely have an associative table (also known as a
junction table or cross-reference table). In our example, this table name might
be `post_topics`.

<!-- prettier-ignore-start -->
[[info | ]]
| You can [jump to this section][self-many-to-many-no-assoc-table] if you don't
| have an associative table in your database.
<!-- prettier-ignore-end -->

The diagram below shows the tables `posts`, `topics`, `post_topics`, and their

> relationship.

Reviewer: ```suggestion
relationships.
```

      primaryKey: true,
    },

    campaignName: {
      sql: `campaign_name`,
      type: `string`,
    },

    campaignId: {
      sql: `campaign_id`,
      type: `number`,
    },
  },
});

cube(`Campaigns`, {
  joins: {
    Transactions: {
      relationship: `hasMany`,
      sql: `${Transactions}.customer_name = ${Campaigns}.customer_name
      AND ${Transactions}.campaign_id = ${Campaigns}.campaign_id`,
    },
  },
});
```

## Directions of joins

The direction of [joins][ref-schema-ref-joins] greatly influences the final
result set. As an example, let's take two cubes, `Orders` and `Customers`:

```javascript
cube('Orders', {
  sql: `SELECT * FROM orders`,

  measures: {
    count: {
      sql: 'id',
      type: 'count',
    },
  },

  dimensions: {
    id: {
      sql: `id`,
      type: `number`,
      primaryKey: true,
    },
  },
});
```

```javascript
cube('Customers', {
  sql: `SELECT * FROM customers`,

  measures: {
    count: {
      sql: 'id',
      type: 'count',
    },
  },

  dimensions: {
    id: {
      sql: `id`,
      type: `number`,
      primaryKey: true,
    },

    customerId: {
      sql: `customer_id`,
      type: `number`,
    },
  },
});
```

> The first case is to calculate total orders revenue. Let's give `totalRevenue`

Reviewer: ```suggestion
The first case is to calculate the total orders revenue. Let's name this metric 
```

> name for this metric. We also need to be aware of the fact that orders can be

Reviewer: ```suggestion
`totalRevenue`. We also need to be aware of the fact that orders can be
```

placed without customer registration (anonymous customers/guest checkouts).
Because of anonymous customers, we should start the join from the `Orders` onto

> the `Customers` cube in order not to lose data about anonymous orders:

Reviewer: ```suggestion
the `Customers` cube so that we do not lose data from anonymous orders:
```

Let's consider an example where we have a many-to-many relationship between
`Users` and `Organizations` through an `OrganizationUsers` cube:

```javascript
cube(`Users`, {

  ...,

  measures: {
    count: {
      type: `count`,
    },
  },

  dimensions: {
    id: {
      sql: `id`,
      type: `string`,
      primaryKey: true,
    },
  },
});

cube(`OrganizationUsers`, {

  ...,

  joins: {
    Users: {
      sql: `${CUBE}.user_id = ${Users}.id`,
      relationship: `hasMany`,
    },
    Organizations: {
      sql: `${CUBE}.organization_id = ${Organizations}.id`,
      relationship: `hasMany`,
    },
  },

  dimensions: {
    id: {
      // Joins require a primary key, so we'll create one on-the-fly
      sql: `CONCAT(${CUBE}.user_id, ':', ${CUBE}.organization_id)`,
      type: `string`,
      primaryKey: true,
    },
  },
});

cube(`Organizations`, {

  ...,

  dimensions: {
    id: {
      sql: `id`,
      type: `string`,
      primaryKey: true,
    },
    name: {
      sql: `category`,
      type: `string`,
    },
  },
});
```

Let's try and execute a query:

```json
{
  "measures": ["Users.count"],
  "dimensions": ["Organizations.name"]
}
```

You'll get an error:

> `Error: Can't find join path to join 'Users', 'Organizations'`. The problem is

Reviewer: ```suggestion
`Error: Can't find join path to join 'Users', 'Organizations'`. The problem is that
```
